File: src/main/java/gtl/spark/java/example/C04/TestLabeledPoint.java
package gtl.spark.java.example.C04;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.mllib.linalg.Vectors;
import org.apache.spark.mllib.regression.LabeledPoint;
import org.apache.spark.mllib.util.MLUtils;

/**
 * 标注点LabeledPoint是一种带有标签（Label/Response）的本地向量，它可以是稠密或者是稀疏的。
 * 在MLlib中，标注点在监督学习算法中被使用。由于标签是用双精度浮点型来存储的，故标注点类型在回归（Regression）和分类（Classification）问题上均可使用。
 * 例如，对于二分类问题，则正样本的标签为1，负样本的标签为0，而对于多类别的分类问题来说，标签则应是一个以0开始的索引序列:0, 1, 2 ...
 * <p>
 * 标注点的实现类是org.apache.spark.mllib.regression.LabeledPoint，请注意它与前面介绍的本地向量不同，并不位于linalg包下
 *
 * 在实际的机器学习问题中，稀疏向量数据是非常常见的，MLlib提供了读取LIBSVM格式数据的支持，该格式被广泛用于LIBSVM、LIBLINEAR等机器学习库。
 * 在该格式下，每一个带标注的样本点由以下格式表示：
 * label index1:value1 index2:value2 index3:value3 ...
 * 其中label是该样本点的标签值，一系列index:value对则代表了该样本向量中所有非零元素的索引和元素值。这里需要特别注意的是，index是以1开始并递增的。
 * MLlib在org.apache.spark.mllib.util.MLUtils工具类中提供了读取LIBSVM格式的方法loadLibSVMFile，其使用非常方便。
 */
public class TestLabeledPoint {
    public static void main(String[] args) {
        // Create a labeled point with a positive label and a dense feature vector.
        LabeledPoint pos = new LabeledPoint(1.0, Vectors.dense(1.0, 0.0, 3.0));

// Create a labeled point with a negative label and a sparse feature vector.
        LabeledPoint neg = new LabeledPoint(0.0, Vectors.sparse(3, new int[]{0, 2}, new double[]{1.0, 3.0}));

        System.out.println(pos);
        System.out.println(neg);

        //MLUtils.loadLibSVMFile
        SparkConf conf = new SparkConf().setAppName("TestLabeledPoint").setMaster("local");
        JavaSparkContext jsc = new JavaSparkContext(conf);
        JavaRDD<LabeledPoint> examples =
                MLUtils.loadLibSVMFile(jsc.sc(), "D:\\devs\\3rdparties\\spark\\spark-2.2.1\\data\\mllib\\sample_libsvm_data.txt").toJavaRDD();
        System.out.println(examples.collect().size());  //100
    }
}

File: src/main/java/gtl/spark/scala/example/C04/TestLabeledPoint.scala
package gtl.spark.scala.example.C04


import org.apache.spark.mllib.linalg.{Vector, Vectors}
import org.apache.spark.mllib.regression.LabeledPoint
import org.apache.spark.mllib.util.MLUtils
import org.apache.spark.{SparkConf, SparkContext}

/**
  * 向量标签对于MLlib中机器学习算法的不同值做标记。例如在分类问题中，可以将不同的数据集分成若干份，以整形数0.1.2...进行标记，即程序
  * 的编写者可以根据自己的需要对数据进行标记
  *
  * LabeledPoint是建立向量标签的静态类，主要有两个方法，features用于显示打印标记点所代表的数据内容，而Label用于显示标记数
  * 除了使用以下两种方法建立向量标签，MLlib还支持直接从数据库中获取固定格式的数据集的方法
  **/
object TestLabeledPoint {
  def main(args: Array[String]) {
    val vd: Vector = Vectors.dense(2, 0, 6) //建立密集向量
    val pos = LabeledPoint(1, vd) //对密集向量建立标记点
    println(pos.features) //打印标记点内容数据     //[2.0,0.0,6.0]
    println(pos.label) //打印既定标记      //1.0
    val vs: Vector = Vectors.sparse(4, Array(0, 1, 2, 3), Array(9, 5, 2, 7)) //建立稀疏向量
    val neg = LabeledPoint(2, vs) //对密集向量建立标记点
    println(neg.features) //打印标记点内容数据   //(4,[0,1,2,3],[9.0,5.0,2.0,7.0])
    println(neg.label) //打印既定标记  //2.0


    // https://yq.aliyun.com/articles/5064
    //http://blog.csdn.net/illbehere/article/details/77162273
    //特征索引不能为0  再因为源代码已经做了减一操作 因此索引只能从1开始
    val conf = new SparkConf().setMaster("local").setAppName("TestLabeledPoint")
    //建立本地环境变量
    val sc = new SparkContext(conf) //建立Spark处理
    val mu = MLUtils.loadLibSVMFile(sc, "D:\\devs\\data\\spark\\D04\\loadLibSVMFile.txt")
    mu.foreach(println) //打印内容
  }
}
